### ISSUE TITLE
Memory leak on unpickling layers

### ISSUE THREAD

**Issue by f0k** [Lasagne/Lasagne#310]
My colleague just stumbled upon what seemed to be a memory leak that eats up his GPU memory. We've condensed it down to a minimum working example:
```python
import sys
import cPickle as pickle
import numpy as np
import theano
import lasagne

VARIANT = int(sys.argv[1]) if len(sys.argv) > 1 else 1

print "Free memory: %g MiB" % (theano.sandbox.cuda.mem_info()[0] / 1024**2)
if VARIANT == 1:
    x = theano.shared(np.ones((10000, 1000), dtype=np.float32) * 42)
elif VARIANT in (2, 3):
    x = lasagne.layers.DenseLayer((None, 10000), 1000,
            W=lasagne.init.Constant(42))
with open('/dev/shm/tmpfile', 'wb') as f:
    pickle.dump(x, f, protocol=-1)
for _ in range(10):
    with open('/dev/shm/tmpfile', 'rb') as f:
        x = pickle.load(f)
    if VARIANT == 3:
        import gc; gc.collect()
    print "Free memory: %g MiB" % (theano.sandbox.cuda.mem_info()[0] / 1024**2)
```

There are three variants:
1. Just pickle and repeatedly unpickle a shared variable
2. Pickle and repeatedly unpickle a `DenseLayer`
3. As 2, but call `gc.collect()` after unpickling

We get the following output:
```
$ python memleak.py 1
Using gpu device 0: GeForce GT 640
Free memory: 522 MiB
Free memory: 483 MiB
Free memory: 483 MiB
Free memory: 483 MiB
Free memory: 483 MiB
Free memory: 483 MiB
Free memory: 483 MiB
Free memory: 483 MiB
Free memory: 483 MiB
Free memory: 483 MiB
Free memory: 483 MiB
$ python memleak.py 2
Using gpu device 0: GeForce GT 640
Free memory: 522 MiB
Free memory: 444 MiB
Free memory: 406 MiB
Free memory: 368 MiB
Free memory: 329 MiB
Free memory: 291 MiB
Free memory: 253 MiB
Free memory: 215 MiB
Free memory: 176 MiB
Free memory: 138 MiB
Free memory: 100 MiB
$ python memleak.py 3
Using gpu device 0: GeForce GT 640
Free memory: 522 MiB
Free memory: 483 MiB
Free memory: 483 MiB
Free memory: 483 MiB
Free memory: 483 MiB
Free memory: 483 MiB
Free memory: 483 MiB
Free memory: 483 MiB
Free memory: 483 MiB
Free memory: 483 MiB
Free memory: 483 MiB
```

So the leak does not occur when explicitly calling the garbage collector, which means that everything seems to become unreachable properly. So technically, it's not a leak. It's still unclear why it doesn't get collected automatically in variant 2 compared to variant 1. Replacing the `DenseLayer` with a very simple dummy class that just creates a shared variable attribute in its constructor doesn't exhibit the leak either. Is there anything we could/should do differently to avoid this problem?

**Comment by benanne** [comment]
The fact that manually calling the garbage collector 'solves' it indicates to me that it isn't really our problem in the first place :P But it would of course be nice if there is something simple we can do to avoid the issue. I guess the only way to find out is to 'deconstruct' the DenseLayer piece by piece until the leak stops occurring.

**Comment by ebenolson** [comment]
It's got something to do with `param` being used as a key in `Layer.params`.

**Comment by f0k** [comment]
Yes, I see it the same way. It's not really our fault, but maybe we're inadvertently doing something that makes it harder for the garbage collector. I don't really know how pickling works, but that should be what we need to investigate.

> It's got something to do with `param` being used as a key in `Layer.params`.

Interesting! So pickling a dictionary with a shared variable as a key is problematic as well?

**Comment by ebenolson** [comment]
I think it's specific to `OrderedDict`

**Comment by ebenolson** [comment]
Minimal example:

    import sys
    import cPickle as pickle
    import numpy as np
    import theano
    import lasagne
    from collections import OrderedDict

    print "Free memory: %g MiB" % (theano.sandbox.cuda.mem_info()[0] / 1024**2)
    x = OrderedDict()
    param = theano.shared(np.ones((10000, 1000), dtype=np.float32) * 42)
    x[param] = True

    with open('/dev/shm/tmpfile', 'wb') as f:
        pickle.dump(x, f, protocol=-1)
    for _ in range(10):
        with open('/dev/shm/tmpfile', 'rb') as f:
            x = pickle.load(f)
        print "Free memory: %g MiB" % (theano.sandbox.cuda.mem_info()[0] / 1024**2)

Output:

    Using gpu device 0: GeForce GTX 770
    Free memory: 4032 MiB
    Free memory: 3955 MiB
    Free memory: 3917 MiB
    Free memory: 3879 MiB
    Free memory: 3840 MiB
    Free memory: 3802 MiB
    Free memory: 3764 MiB
    Free memory: 3726 MiB
    Free memory: 3687 MiB
    Free memory: 3649 MiB
    Free memory: 3611 MiB

**Comment by f0k** [comment]
Yep, that nailed it! Good work, Sherlock! We don't even need pickling:
```python
import numpy as np
import theano
from collections import OrderedDict

def make_shared():
	return theano.shared(np.ones((10000,1000), dtype=np.float32) * 42)

print "Free memory: %g MiB" % (theano.sandbox.cuda.mem_info()[0] / 1024**2)
for _ in range(10):
	x = OrderedDict({make_shared(): 'foo'})
	print "Free memory: %g MiB" % (theano.sandbox.cuda.mem_info()[0] / 1024**2)
```
This means we're even subject to the leak if we just create and throw away layers without manually invoking the garbage collector in between. This seems like a serious gotcha.

Using a shared variable as a value doesn't have the same problem. So... we could change `self.params` to be a dictionary mapping tags to shared variables instead of vice versa, but that would be silly (it makes it more difficult to inspect and modify, and helping the garbage collector is not a valid reason to change the API). We could also use some replacement for `OrderedDict`, or try to file a bug report against Python (assuming that *anything* used as an `OrderedDict` key doesn't get collected properly, otherwise it may be Theano's business).

**Comment by f0k** [comment]
It's not a fault of Theano:
```python
from collections import OrderedDict
import gc

class myobject():
    def __del__(self):
        print "Deleting %r" % self

print "Ordinary dict without manual gc:"
for _ in range(3):
    print "Creating..."
    {myobject(): 'foo'}

print "OrderedDict with manual gc:"
for _ in range(3):
    print "Creating..."
    OrderedDict([(myobject(), 'foo')])
    gc.collect()

print "OrderedDict without manual gc:"
for _ in range(3):
    print "Creating..."
    OrderedDict([(myobject(), 'foo')])
```
gives:
```
jan@urlaub:/dev/shm$ python memleak3.py
Ordinary dict without manual gc:
Creating...
Deleting <__main__.myobject instance at 0x7f9d49ecaea8>
Creating...
Deleting <__main__.myobject instance at 0x7f9d49ecaea8>
Creating...
Deleting <__main__.myobject instance at 0x7f9d49ecaea8>
OrderedDict with manual gc:
Creating...
Deleting <__main__.myobject instance at 0x7f9d49ecaea8>
Creating...
Deleting <__main__.myobject instance at 0x7f9d49f12a28>
Creating...
Deleting <__main__.myobject instance at 0x7f9d49f12758>
OrderedDict without manual gc:
Creating...
Creating...
Creating...
Deleting <__main__.myobject instance at 0x7f9d49f12950>
Deleting <__main__.myobject instance at 0x7f9d49f12710>
Deleting <__main__.myobject instance at 0x7f9d49f129e0>
```

**Comment by ebenolson** [comment]
Seems to be fixed in python 3.4

**Comment by f0k** [comment]
We could also roll our own `OrderedDict` that uses a combination of a `dict` and a `list`, since we don't need O(1) deletion anyway. But let's first wait if others stumble upon the same issue, I don't feel good about complicating our code just to work around an internal Python problem that surfaces in rare circumstances.

**Comment by f0k** [comment]
Ahaa, found another solution: Calling `clear()` on the dictionary removes all items from the doubly-linked list so they can be trivially garbage-collected (and are garbage-collected). We could just add the following to the `Layer` base class:
```python
def __del__(self):
    # allow immediate collection of parameters (Lasagne Issue #130)
    self.params.clear()
```
Any oppositions?

**Comment by benanne** [comment]
I'm okay with adding this to the `Layer` class, since that's more or less hidden from the user and if we have a comment in there (like in your example) it shouldn't confuse people looking at the source too much. I think subclassing `OrderedDict` is probablhy a step too far though. Documenting this shortcoming is probably better.

**Comment by f0k** [comment]
As Eben pointed out, the issue is fixed with Python 3.4, though. I've just compared the source and they avoid the cyclic references by using weakrefs for the backward direction in the doubly-linked list. So we should probably use the deletion trick only for Python versions before that, using a `sys.version` check. This will make it look even more silly, but I think the problem is important enough -- creating multiple models within a single process is not too uncommon of a use case (for parameter studies / grid searches). I'll do a PR tomorrow.

**Comment by benanne** [comment]
Bummer! I don't know if a note in the docs suffices though... this sounds like something that could bite people without them even realising it ("I'm running out of memory, my model must be too big for my GPU."). And it's probably pretty challenging to track down for people who are not Python wizards :)

What is a use case where a cycle of `Layer` instances would come up though? Aren't DAGs the only possibility?

**Comment by ebenolson** [comment]
So, an alternative way to fix this would be to modify `Layer` not to use `OrderedDict`. We could instead have `Layer.params = []` and `Layer.param_tags = {}`, using the list to maintain the ordering. 

It would be a pretty minimal change, and might even be clearer than currently (I wouldn't expect `Layer.params` to have anything to do with tags without reading the code).

**Comment by ebenolson** [comment]
Oh, I see @f0k already mentioned that method earlier.

**Comment by f0k** [comment]
> I don't know if a note in the docs suffices though... this sounds like something that could bite people without them even realising it

Yes, I think merely documenting it is not the solution.

> What is a use case where a cycle of `Layer` instances would come up though?

None so far. You need to tinker with `Layer.input_layer` directly to construct a cycle. It just occurred to me that introducing a `__del__` method has some theoretical disadvantage.

> We could instead have `Layer.params = []` and `Layer.param_tags = {}`, using the list to maintain the ordering.
> Oh, I see @f0k already mentioned that method earlier.

Not directly, I was thinking about an own `OrderedDict` class that uses a `list` instead of the doubly linked list. (The `weakref` implementation of Python 3.4 is arguably more efficient.)

> It would be a pretty minimal change, and might even be clearer than currently (I wouldn't expect `Layer.params` to have anything to do with tags without reading the code).

Yes, it would be cleaner, but it would also make some operations more cumbersome. Disabling the bias in a layer after constructing it can currently be achieved by:
```python
del layer.params[layer.b]
layer.b = None
```
It would become:
```python
layer.params.remove(layer.b)
del layer.param_tags[layer.b]
layer.b = None
```
Collecting all the parameters of another layer to make them your own can currently be achieved by:
```python
self.params.update(other.params)
```
It would become:
```python
self.params.extend(other.params)
self.param_tags.update(other.param_tags)
```
So you'd always need to ensure the two stay in sync. I'm not really opposed to it, though. `Layer.get_params()` wouldn't be a lot different from what it's now. Having `Layer.params` be a `list` would probably be less surprising than the `OrderedDict`, but it may lead to `Layer.param_tags` being overseen.

**Comment by nouiz** [comment]
What about adding in Theano a new OrderedDict that have the fix python 3.4
have? We already have one for python 2.6. But it probably have the same
problem.

Just an idea, I don't have time to work on it and I'm not sure of all the
consequence of this idea.

**Comment by f0k** [comment]
Hey Fred, nice to see you're watching :)

> What about adding in Theano a new OrderedDict that have the fix python 3.4 have?

Yeah, I had the same thought later, because Theano's update dictionaries are affected in the same way, and Lasagne is dependent on Theano already. This could be a good alternative.

> Just an idea, I don't have time to work on it and I'm not sure of all the consequence of this idea.

The major hurdle is that people will need to change their existing code to use Theano's implementation instead of the one in Python <3.4 if they want to avoid this "leak", and nobody will know that. But we can at least easily change `lasagne.layers.Layer.params` and the `lasagne.updates` module.

----

Just to get things straight, it's not actually a memory leak. In Python, each object has a reference counter that tracks how many references to the object exist. Whenever this reaches zero, it is deleted immediately (without any intervention by the garbage collector).
However, the keys of an `OrderedDict` are entered into a doubly-linked list, and thus become part of a reference cycle (before Python 3.4). This means that even when the `OrderedDict` is deleted (so nobody has any references to the doubly-linked list), the list nodes still have a positive refcount and are not deleted immediately, so the keys stay alive as well. The only way such objects can be freed is via the garbage collector, which searches for reference cycles that are unreachable from any "active" objects. It is invoked in regular intervals, so in theory, we shouldn't be affected by the `OrderedDict` issue anyway. But in practice, it seems it's involved much too seldom: My colleague trained a model for 100 epochs between each unpickling call and ran out of GPU memory after 3 hours. So we need to do *something*.

----

We've got the following alternatives now:

1. ~~Just document to invoke `gc.collect()` after deleting a model~~
2. Replace `Layer.params` by `Layer.params = []` and `Layer.param_tags = {}`
3. Replace `Layer.params` and the update dictionaries in `lasagne.updates.*` by a custom `OrderedDict` implementation from `theano`

Number one will only help a miniscule fraction of users.
Number two makes things a bit cleaner and easier to understand, but also [a bit harder to manage](https://github.com/Lasagne/Lasagne/issues/310#issuecomment-114657247) (because nothing enforces `Layer.param_tags.keys()` to stay in sync with `Layer.params` like the `OrderedDict` naturally does). It also doesn't fully solve the problem, because any parameters of the model that are part of an update dictionary will still be kept alive indefinitely. And finally, it would mean another change of the API (`Layers.params` is publicly visible and must be considered part of the API -- e.g., I'd have to change the [batch normalization implementation](https://gist.github.com/f0k/f1a6bd3c8585c400c190) again).
Number three transparently resolves the issue for all Lasagne users except those that build a custom updates dictionary from scratch, starting with an `OrderedDict`.

So I'd vote for number three, with number two still open for discussion (not for solving this issue, but as an alternative API).

**Comment by benanne** [comment]
3 sounds like the way to go. I'm not a big fan of 2, sounds like it would make it easy to forget about one or the other. But if 3 doesn't happen I guess 2 is the next best option.

**Comment by f0k** [comment]
> I'm not a big fan of 2, sounds like it would make it easy to forget about one or the other.

Yes, that's what's elegant about the `OrderedDict`. I'm torn between that argument and the fact that separating `params` and `param_tags` seems so much clearer. If we go for 2, `get_params()` should gracefully handle something being entered in `params` and not in `param_tags`.

> But if 3 doesn't happen I guess 2 is the next best option.

Actually, 2 is fully independent of 3, because it's not a solution to the same problem (sure, it will avoid the leak in `Layer.params`, but it won't fix it in `lasagne.updates.*`, so it doesn't help much).

**Comment by benanne** [comment]
I don't really understand why separating them would seem clearer. If anything it results in more code and more bookkeeping. I guess that's a subjective thing though :) are there any concrete advantages to separating them?

**Comment by f0k** [comment]
> I don't really understand why separating them would seem clearer.

Just conceptually. I think it's easier to understand "this layer has some parameters (`params`), and then it has tags for these parameters (`param_tags`)" than "this layer has some parameters, and they also have tags, but we just call all that `params` anyway".

> are there any concrete advantages to separating them?

Technically not so many, no. It's easier to collect all the parameters when you're not filtering them (you can just collect all `Layer.params` lists instead of `Layer.params.keys()`, the latter of which has to copy the doubly linked list into a list in case of Python 2). But there's no crucial technical advantage, just the conceptual one.

**Comment by ebenolson** [comment]
What about: 4. Make `Layer.params` a list, store tags as a list on each `param`.

When I suggested this before, I think the main objection was that having params shared across layers might cause tags to be overwritten, but I think we an equivalent situation currently: if different layers have contradictory tags for a shared param, it will give confusing results when trying to filter `get_all_params`.

**Comment by f0k** [comment]
> What about: 4. Make `Layer.params` a list, store tags as a list on each `param`.

(A `set` would be better for the latter, we don't need them to be ordered.)

> I think the main objection was that having params shared across layers might cause tags to be overwritten, but I think we an equivalent situation currently: if different layers have contradictory tags for a shared param, it will give confusing results when trying to filter `get_all_params`.

The objection is also about parameters shared across different networks (e.g., when taking a parameter of an existing network and using it to create a layer in a different network). Another objection was that we explicitly want to tag our parameters, not the Theano graph (unlike `blocks`), so it would be good to have the tagging encapsulated in the `Layer` class and not touch the expressions constituting the graph. However, an advantage of tying the tags to the shared parameters would be that you could still easily obtain the tags of each parameter after collecting them via `get_all_params`.

In any case, this discussion probably belongs into #164 instead of here, as we need a different `OrderedDict` implementation regardless of what we do with `Layer.params`.

**Comment by nouiz** [comment]
I don't have time now to work on a new ordereddict. Do one of you can
search on the web if a good implementation already exist? If so, it would
be easy to move it in theano (modulo the license)

**Comment by f0k** [comment]
> I don't have time now to work on a new ordereddict. Do one of you can search on the web if a good implementation already exist?

The most sane way would be to update [Theano's copy](https://github.com/Theano/Theano/blob/master/theano/compat/python2x.py) of the MIT-licensed [ordereddict backport on PyPI](https://pypi.python.org/pypi/ordereddict) to use `weakref.proxy` for the backward links, as in Python 3.4, and possibly contribute the change back to PyPI (I've already asked its maintainer @sah if he's interested, but haven't heard back yet).
The backport is almost the same as the code in Python 2.7 (which isn't that surprising as they're both by Raymond Hettinger: http://bugs.python.org/issue5397). I'm not sure if we'd run into license issues by just porting the `weakref.proxy` idea from the Python 3.4 codebase to the backport? There's hardly any design choices anyway... if we want to use weak references, then `weakref.proxy` is the easiest way, and we need to turn the linked-list nodes from `list` instances into objects because `weakref` does not support `list`. That's exactly what has happened in Python 3.4.

**Comment by nouiz** [comment]
I think it is ok for the lisence to use the same idea and as there is not
many choise, if the code end up similar, it is fine. But for license, I
think we can't copy the code.
